Feature/disjoint geo new

vissarion opened the pull request:
**Title:** Algorithm for disjoint predicate of a segment and a box in spherical and geographic coordinate systems. 

_Main idea:_ The algorithm uses an angle test and if it is not enough to judge computes the vertex of the segment (point of maximum latitude) and test if it is covered by the box. 

_Main contributions_
- A formula for computing the longitude of a segment's vertex is implemented. 
- The segment envelope methods are slightly changed to provide more efficient versions needed by disjoint segment box algorithm. 
- New strategies segment_box_spherical and segment_box_geographic are implemented to handle spherical or geographical geometries respectively.

barendgehrels commented on include/boost/geometry/algorithms/detail/disjoint/segment_box.hpp:
```diff
@@ -36,10 +38,194 @@ namespace boost { namespace geometry
 {
 
+template <typename CT, typename CS_Tag>
+struct disjoint_segment_box_call_vertex_longitude{
```
Can you put the curly brace one line lower? (here and below)

barendgehrels commented on include/boost/geometry/algorithms/detail/disjoint/segment_box.hpp:
```diff
@@ -36,10 +38,194 @@ namespace boost { namespace geometry
+                           CT vertex_lat,
+                           CT,
+                           CT alp1,
+                           Strategy azimuth_strategy)
```
Should be passed const& I suppose

barendgehrels commented on include/boost/geometry/algorithms/detail/disjoint/segment_box.hpp:
```diff
@@ -36,10 +38,194 @@ namespace boost { namespace geometry
+                                        CT lat2,
+                                        CT vertex_lat,
+                                        CT alp1,
+                                        Strategy azimuth_strategy)
```
Same here, const& (I mean mainly the strategy)

barendgehrels commented on include/boost/geometry/algorithms/detail/disjoint/segment_box.hpp:
```diff
@@ -36,10 +38,194 @@ namespace boost { namespace geometry
+
+        segment_point_type p0, p1;
+        geometry::detail::assign_point_from_index<0>(segment, p0);
+        geometry::detail::assign_point_from_index<1>(segment, p1);
```
These two calls need an extra include

barendgehrels commented on include/boost/geometry/algorithms/detail/disjoint/segment_box.hpp:
```diff
@@ -36,10 +38,194 @@ namespace boost { namespace geometry
+        CT b_lon_min = geometry::get<geometry::min_corner, 0>(box) * math::d2r<CT>();
+        CT b_lat_min = geometry::get<geometry::min_corner, 1>(box) * math::d2r<CT>();
+        CT b_lon_max = geometry::get<geometry::max_corner, 0>(box) * math::d2r<CT>();
+        CT b_lat_max = geometry::get<geometry::max_corner, 1>(box) * math::d2r<CT>();
```
Ah, we've not yet a get_as_radian here... This is actually not completely right (I think) because you assume coordinates are in degrees, which is not required for this algorithm.

barendgehrels commented on include/boost/geometry/algorithms/detail/disjoint/segment_box.hpp:
```diff
@@ -36,10 +38,194 @@ namespace boost { namespace geometry
+                                           azimuth_strategy);
+
+        segment_point_type p_vertex_deg(vertex_lon * math::r2d<CT>(),
+                                        vertex_lat * math::r2d<CT>());
```
same here

barendgehrels commented on include/boost/geometry/algorithms/detail/envelope/segment.hpp:
```diff
@@ -300,6 +289,83 @@ class envelope_segment_impl
+              typename CalculationType,
+              typename Box,
+              typename Strategy
+            >
```
We ususally have the layout different for template arguments (i.e. same as for structs {} )

barendgehrels commented on include/boost/geometry/formulas/vertex_longitude.hpp:
```diff
@@ -0,0 +1,174 @@
+                           T const lat2,
+                           T const lat3,
+                           T const dlon)
+    {
```
I think (here and in other places) this should be T const&, for user defined numeric types.

barendgehrels commented on include/boost/geometry/formulas/vertex_longitude.hpp:
```diff
@@ -0,0 +1,174 @@
+
+        CT const sin_alp1 = sin(alp1);
+        CT const sin_alp0 = sin_alp1 * cos_bet1;
+        CT const sin_alp2 = sin_alp1 * cos_bet1 / cos_bet2;
```
Are we sure cos_bet2 != 0 ?

barendgehrels commented on include/boost/geometry/strategies/geographic/disjoint_segment_box.hpp:
```diff
@@ -0,0 +1,132 @@
+    typename Spheroid = geometry::srs::spheroid<CalculationType>,
+    template <typename, bool, bool, bool, bool, bool> class Inverse =
+        geometry::formula::thomas_inverse
+>
```
So this layout is OK

barendgehrels commented on include/boost/geometry/strategies/geographic/disjoint_segment_box.hpp:
```diff
@@ -0,0 +1,132 @@
+        typedef typename point_in_geometry_strategy<Segment, Box>::type strategy_type;
+
+        return strategy_type();
+    }
```
This relates to Adam's review about getting strategies

awulkiew commented on include/boost/geometry/algorithms/detail/disjoint/segment_box.hpp:
```diff
@@ -36,10 +38,194 @@ namespace boost { namespace geometry
+        azimuth_strategy.apply(lon1, lat1, b_lon_min, b_lat_max, a_b2);
+        azimuth_strategy.apply(lon1, lat1, b_lon_max, b_lat_max, a_b3);
+
+        bool b0(alp1 > a_b0), b1(alp1 > a_b1), b2(alp1 > a_b2), b3(alp1 > a_b3);
```
This looks unusual, using assignment operator would be more clear. Plus typically initializations are put in separate lines in BG, no matter how short they are.

awulkiew commented on include/boost/geometry/algorithms/detail/disjoint/segment_box.hpp:
```diff
@@ -36,10 +38,194 @@ namespace boost { namespace geometry
+
+        // if the box is above (below) the segment in northern (southern)
+        // hemisphere respectively then there is not intersection
+        if (~(b0 & b1 & b2 & b3) & (b0 | b1 | b2 | b3))
```
Why bitwise operators (after implicit conversion to int) are used here, not logical operators? AFAIU the result would be the same and the intention would be more clear. Or is there some reason?

vissarion commented:
Thanks for the reviews! I addressed your comments. Mainly regarding degree/radian issue. I created get_as_radian functions for boxes (and segments). Now disjoint works for both degree and radian input coordinate systems. However, it does not work correctly for a mixed input (e.g. box in degrees and segment in radians) due to covered_by function. This can be resolved in a future commit if we need it as functionality.

awulkiew commented on include/boost/geometry/core/radian_access_box_segment.hpp:
```diff
@@ -0,0 +1,151 @@
+    e.g. spherical or geographic coordinate systems
+*/
+template <std::size_t Index, std::size_t Dimension, typename Geometry>
+inline typename fp_coordinate_type<Geometry>::type get_as_radian(Geometry const& geometry)
```
I would put it in one file with `get_as_radian` for points. It'd be simpler for including and consistent with already implemented `get`.

awulkiew commented:
Thanks, I'm ok with merging. @barendgehrels?

awulkiew commented:
@barendgehrels are you ok with merging this PR?

barendgehrels commented:
Sure, I'm OK, my points are addressed.
